### PR TITLE
Add CustomDebugStringConvertible and CustomStringConvertible conformance

### LINKED ISSUE #50: Add CustomStringConvertible and CustomDebugStringConvertible conformance
https://developer.apple.com/library/prerelease/ios/documentation/Swift/Reference/Swift_CustomDebugStringConvertible_Protocol/index.html#//apple_ref/swift/intfp/CustomDebugStringConvertible/s:vPSs28CustomDebugStringConvertible16debugDescriptionSS

### PR DESCRIPTION
**Pull request by JonathanHoffman**
Progress! 

TODO: 
- [x] implement `CustomStringConvertible` and `CustomDebugStringConvertible` to the Expression types.
- [x] write a playground page exploring these types, their strings, and their developer friendliness. (what would you like to see in such a page?) 

Closes #50

### CODE REVIEW

File: DiceKit/Die.swift [hunk 1]
```diff
@@ -116,3 +116,14 @@ extension Die: ExpressionType {
 public func == (lhs: Die, rhs: Die) -> Bool {
     return lhs.sides == rhs.sides
 }
+
+// MARK: - Debug
```
Comment (brentleyjones) [inline DiceKit/Die.swift]: `CustomDebugStringConvertible`. When we conform to a protocol we state the name as is in the section.

Comment (brentleyjones) [inline DiceKit/Die.swift]: Also, should be a new line after it. We only don't do newlines in the tests, because there we are naming the extension as well. If anything, we should use newlines there as well, just pointing out why there is a difference.

File: DiceKit/Die.swift [hunk 2]
```diff
@@ -116,3 +116,14 @@ extension Die: ExpressionType {
...
+extension Die: CustomDebugStringConvertible {
+    
+    public var debugDescription: String {
+        get {
+            return "I am a Die! I have \(self.sides) sides!"
```
Comment (brentleyjones) [inline DiceKit/Die.swift]: What we should show is whatever you would expect to show on the right hand side in a playground. Let's see what some defaults are. `po` in the debugger gives a nice looking view:

```
(lldb) po Die(sides: 6)
▿ DiceKit.Die
  - sides : 6
```

And this is default `debugPrint`:

```
(lldb) po debugPrint(Die())
DiceKit.Die(sides: 6)
```

I don't know about you, but `DiceKit.Die(sides: 6)` is very verbose. I think something like `Die(6)` might be better debugPrint.

Where it will get really interesting is with expressions. Those we will want to basically reverse into operators.

Comment (brentleyjones) [inline DiceKit/Die.swift]: Let's also look into `print` versus `debugPrint`. I think `print` should show `d6` and `8`, while `debugPrint` should show `Die(sides: 6)` and `Constant(8)`?

Comment (JonathanHoffman) [inline DiceKit/Die.swift]: Yeah. I have some code right now that will show expression components separated by their operator, and d notation for the die. But for debug I think it might be better if the type was more clear. 

Also has me thinking about some stuff like "if the multiplier of a MultiplicationExpression is a Constant, print it like this. Else, print it in ()" etc. Then we would practically have reverse string code! :stuck_out_tongue:

**Comment by brentleyjones** [comment]
> I find its nice to make a playground to explore these.

How about adding a page that showcases how dev friendly our types are? The first section can be on these two conformances. Later we can add a QuickLook section?

File: DiceKit/Die.Roll.swift [hunk 3]
```diff
@@ -35,3 +35,27 @@ extension Die.Roll: ExpressionResultType {
...
+    
+}
+
+// MARK: - CustomStringConvertible
+
+extension Die.Roll: CustomStringConvertible {
+    
+    public var description: String {
+        get {
+            return "\(die) roll"
```
Comment (brentleyjones) [inline DiceKit/Die.Roll.swift]: These should definitely show the result of the roll.

File: DiceKit/MultiplicationExpressionResult.swift [hunk 4]
```diff
@@ -39,3 +39,27 @@ public struct MultiplicationExpressionResult<MultiplierResult: protocol<Expressi
...
+    
+}
+
+// MARK: - CustomStringConvertible
+
+extension MultiplicationExpressionResult: CustomStringConvertible {
+    
+    public var description: String {
+        get {
+            return "\(multiplierResult) * \(multiplicandResults)"
```
Comment (brentleyjones) [inline DiceKit/MultiplicationExpressionResult.swift]: So, this one is incorrect. The `multiplierResult` was used to generate all of the `multiplicandResults`. I believe that here you just show all of them being added together: `(5 +  3 + 1 + 5 + 2)`

**Comment by brentleyjones** [comment]
I don't think there is much boilerplate at all. It's just the way they work. So, overall, :+1:.

File: DiceKit/MultiplicationExpressionResult.swift (hunk 4 above)
Comment (JonathanHoffman) [inline DiceKit/MultiplicationExpressionResult.swift]: Ah, I clearly didn't check to make sure this one made sense. I'll fix that in my next push, which should be the one ready for final review.

**Comment by JonathanHoffman** [comment]
I really like the idea of writing a page for this! I can cover how dev friendly everything is while giving an idea of why each type exists and how to use them. It will also help people understand what to expect from each type without reading all the tests (see my MultiplicationExpressionResult goof).

File: DiceKit/AdditionExpression.swift [hunk 5]
```diff
@@ -72,3 +72,27 @@ public func + <R: protocol<ExpressionType, Equatable>>(lhs: Int, rhs: R) -> Addi
 public func + <L: protocol<ExpressionType, Equatable>>(lhs: L, rhs: Int) -> AdditionExpression<L, Constant> {
     return AdditionExpression(lhs, Constant(rhs))
 }
+
+// MARK: - CustomDebugStringConvertible
```
Comment (brentleyjones) [inline DiceKit/AdditionExpression.swift]: I think we should list `CustomStringConvertible` before `CustomDebugStringConvertible`, and both of them right before the `Equatable` section.

File: DiceKit/Die.Roll.swift [hunk 6]
```diff
@@ -35,3 +35,27 @@ extension Die.Roll: ExpressionResultType {
 public func == (lhs: Die.Roll, rhs: Die.Roll) -> Bool {
     return lhs.die == rhs.die && lhs.value == rhs.value
 }
+
+// MARK: - CustomDebugStringConvertible
+
+extension Die.Roll: CustomDebugStringConvertible {
+    
+    public var debugDescription: String {
+        get {
```
Comment (brentleyjones) [inline DiceKit/Die.Roll.swift]: When only a getter, we don't explicitly use `get`. Just `return` instead.

File: DiceKit/Die.swift [hunk 7]
```diff
@@ -134,3 +134,27 @@ extension Die: ExpressionType {
...
+    
+}
+
+// MARK: - CustomStringConvertible
+
+extension Die: CustomStringConvertible {
+    
+    public var description: String {
+        get {
+            return "d\(sides)"
```
Comment (brentleyjones) [inline DiceKit/Die.swift]: If negative should this instead be `d(-12)`? I think it would look better for that case.

File: DiceKit/MultiplicationExpressionResult.swift [hunk 8]
```diff
@@ -45,3 +45,40 @@ extension MultiplicationExpressionResult: ExpressionResultType {
...
+
+extension MultiplicationExpressionResult: CustomDebugStringConvertible {
+    
+    public var debugDescription: String {
+        get {
+            let multiplierResultDebugString = String(reflecting: multiplierResult)
+            let multiplicandDebugStrings = multiplicandResults.map { String(reflecting: $0) }
+            let multiplicandResultsDebugDescription = " + ".join(multiplicandDebugStrings)
+            if negateMultiplicandResults {
+                return "((\(multiplierResultDebugString)) *> -(\(multiplicandResultsDebugDescription)))"
```
Comment (brentleyjones) [inline DiceKit/MultiplicationExpressionResult.swift]: I would extract this to be:

```
let multiplicandResultsSign = negateMultiplicandResults ? "-" : ""
return "((\(multiplierResultDebugString)) *> \(sign)(\(multiplicandResultsDebugDescription)))"
```

Then less duplication.

File: DiceKit/MultiplicationExpressionResult.swift [hunk 9]
```diff
@@ -45,3 +45,40 @@ extension MultiplicationExpressionResult: ExpressionResultType {
...
+// MARK: - CustomStringConvertible
+
+extension MultiplicationExpressionResult: CustomStringConvertible {
+    
+    public var description: String {
+        get {
+            let multiplicandStrings = multiplicandResults.map { String($0) }
+            let multiplicandResultsDescription = " + ".join(multiplicandStrings)
+            if negateMultiplicandResults {
+                return "((\(multiplierResult)) *> -(\(multiplicandResultsDescription)))"
```
Comment (brentleyjones) [inline DiceKit/MultiplicationExpressionResult.swift]: Same `sign` comment.

File: DiceKitTests/AdditionExpressionResult_Tests.swift [hunk 10]
```diff
@@ -80,3 +80,32 @@ extension AdditionExpressionResult_Tests {
...
+    func test_CustomStringConvertible() {
+        let expression = d(10) + 5
+        let evaluation = expression.evaluate()
+        let expected = "\(evaluation.leftAddendResult.value)|10 + 5"
+        
+        let result = String(evaluation)
+        
+        expect(result) == expected
+    }
+}
```
Comment (brentleyjones) [inline DiceKitTests/AdditionExpressionResult_Tests.swift]: Newline

File: DiceKitTests/AdditionExpressionResult_Tests.swift [hunk 11]
```diff
@@ -80,3 +80,32 @@ extension AdditionExpressionResult_Tests {
...
+    
+    func test_CustomStringConvertible() {
+        let expression = d(10) + 5
+        let evaluation = expression.evaluate()
+        let expected = "\(evaluation.leftAddendResult.value)|10 + 5"
+        
+        let result = String(evaluation)
+        
+        expect(result) == expected
+    }
```
Comment (brentleyjones) [inline DiceKitTests/AdditionExpressionResult_Tests.swift]: Newline

File: DiceKitTests/MultiplicationExpressionResult_Tests.swift [hunk 12]
```diff
@@ -121,3 +121,34 @@ extension MultiplicationExpressionResult_Tests {
...
+        let evaluation = expression.evaluate()
+        let expected = "((2) *> (\(evaluation.multiplicandResults[0]) + \(evaluation.multiplicandResults[1])))"
+        
+        let result = String(evaluation)
+        
+        expect(result) == expected
+    }
+    
+}
+
```
Comment (brentleyjones) [inline DiceKitTests/MultiplicationExpressionResult_Tests.swift]: Newline and weird spaces

File: DiceKitTests/NegationExpression_Tests.swift [hunk 13]
```diff
@@ -106,3 +106,31 @@ extension NegationExpression_Tests {
...
+    
+    func test_CustomStringConvertible() {
+        let expression = -(d(8))
+        let expected = "-d8"
+        
+        let result = String(expression)
+        
+        expect(result) == expected
+    }
+}
```
Comment (brentleyjones) [inline DiceKitTests/NegationExpression_Tests.swift]: Newline

**Comment by brentleyjones** [comment]
A couple small things, but overall I like it. :sparkles:

File: DiceKit/Die.Roll.swift (hunk 6 above)
Comment (JonathanHoffman) [inline DiceKit/Die.Roll.swift]: Noted for future.

File: DiceKit/Die.swift (hunk 7 above)
Comment (JonathanHoffman) [inline DiceKit/Die.swift]: Yeah, definitely.

File: DiceKit/MultiplicationExpressionResult.swift (hunk 8 above)
Comment (JonathanHoffman) [inline DiceKit/MultiplicationExpressionResult.swift]: Nice, I like that. Always looking for ways to be more DRY.

File: DiceKitTests/MultiplicationExpressionResult_Tests.swift (hunk 12 above)
Comment (JonathanHoffman) [inline DiceKitTests/MultiplicationExpressionResult_Tests.swift]: I don't see this one?

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 14]
```diff
@@ -0,0 +1,50 @@
+//: ## Exploring Expressions
+//: 
+//: Our expressions are more robust than what is needed for typical tabletop play. In fact, our generic expression system is capable of modeling expressions that would otherwise be considered too cumbersome or complex to physically roll. In order to help developers understand what their expressions actually represent, we have added various tools to make working with expressions as easy as possible.
+
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: We have the comments sit on top of the expressions in the playground. So this would be removed.

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+import DiceKit
+
+//: Lets start with the negative-sided die.
+
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: As well as this.

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+let negativeDie = d(-12)
+
+//: A die with negative sides has results in the range `-sides...-1`. We have two ways to represent an expression as a string.
+
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: Etc.

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+//: 
+//: Our expressions are more robust than what is needed for typical tabletop play. In fact, our generic expression system is capable of modeling expressions that would otherwise be considered too cumbersome or complex to physically roll. In order to help developers understand what their expressions actually represent, we have added various tools to make working with expressions as easy as possible.
+
+import DiceKit
+
+//: Lets start with the negative-sided die.
+
+let negativeDie = d(-12)
+
+//: A die with negative sides has results in the range `-sides...-1`. We have two ways to represent an expression as a string.
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: "A die with negative sides **by default** has results in the range `-sides...-1`."

Add "by default", and maybe reference that we have multiple ways to setup the roller? Maybe do that in the `Die` page, and just reference it here?

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+//: `oneD2` is of the type `MultiplicationExpression<Constant, Die>` because it is made by multiplying a `Consant` (the expression type `Int` is converted to) by a `Die`.
+
+oneD2 is MultiplicationExpression<Constant, Die>
+
+//: We can make a new multiplication expression with these two expressions. 
+
+let oneD2d3 = oneD2 * d3
+
+//: This expression works like a "normal" multiplication expression, but the multiplier is one half the time, and two the other half of the time. When we `evaluate()` a multiplication expression, we can see what each side of the expression evaluated to. First, an easy case:
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: " but the multiplier is one half the time, and two the other half"

Let's say `1` and `2` instead, so they read easier. I had to re-read this a couple times to understand "one half" wasn't "1/2".

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+let twoD20 = 2 * d(20)
+let twoD20result = twoD20.evaluate()
+
+//: This result type has a `.multiplierResult` and  .`multiplicandResults`. The multiplier is always 2, because it is a constant. The multiplicand is evaluated and summed n number of times to give the result, where n is the multiplier result. We seperate the multiplier and multiplicand with the symbol `*>` in the description, which says that the result of the left hand side is the length of the right hand side.
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: Should be `.multiplicandResults`, the period is on the wrong side.

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+let oneD2d3result = oneD2d3.evaluate()
+
+//: Now we see the result of `oneD2d3`, which is made up of three terms sperated by `*>`. The leftmost term is the multiplier for 1d2. It must always be 1. The middle term is either a 1|2 or a 2|2, depending on what the d2 rolled. Finally, the rightmost term is either a single d3 roll or the sum of two d3 rolls, depending on what the d2 rolled. The result of the expression is always the sum of the rightmost term.
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: Normally when talking about an expression we have the comment above it. I think this needs to be above `oneD2d3result`?

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+let value = oneD2d3result.value
+
+//: Try making more expressions in this playground. There is no limitation on what expression types you can put together to make a new expression. If your expression gets too complicated to easily tell its type, use the `dump()` function to investigate it in the console window.
+
+//: [Next](@next)
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift:41]: This should be previous. And on Probability you should add next.

**Comment by brentleyjones** [comment]
I really liked the playground. Great job! :sparkles:

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift (hunk 19 above)
Comment (JonathanHoffman) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: Lol, I had to read this a few times to see it.

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift (hunk 20 above)
Comment (JonathanHoffman) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift]: Yeah, this one is weird because I just plopped a new line without talking about it first. But elsewhere on this page I did save remarks for after a line. So I went through and made more changes to be consistent. I added as a new commit instead of amending so that the style changes can be compared.

File: TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift (hunk 21 above)
Comment (JonathanHoffman) [inline TomeOfKnowledge.playground/Pages/Exploring Expressions.xcplaygroundpage/Contents.swift:41]: oops. I wanted this to come right after expressions, but my playground was acting buggy when I looked at it last time. I will try to amend and fix now.

File: TomeOfKnowledge.playground/contents.xcplayground [hunk 22]
```diff
@@ -1,8 +1,9 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<playground version='6.0' target-platform='osx' requires-full-environment='true' display-mode='rendered'>
+<playground version='6.0' target-platform='osx' requires-full-environment='true' display-mode='raw'>
```
Comment (brentleyjones) [inline TomeOfKnowledge.playground/contents.xcplayground]: Revert this change, since we want it to be rendered.

**Comment by brentleyjones** [comment]
Just the one revert, then good to merge :+1:

File: TomeOfKnowledge.playground/contents.xcplayground (hunk 22 above)
Comment (JonathanHoffman) [inline TomeOfKnowledge.playground/contents.xcplayground]: Hmm... not sure what this is? I'll take a peek at it. Oh, that must be because I committed with the raw markup viewed?

Comment (brentleyjones) [inline TomeOfKnowledge.playground/contents.xcplayground]: Yep.
